[DR-2969] Display DUOS group's last synced time

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-2969

If a snapshot is linked to a DUOS dataset, we now display the DUOS Firecloud's last synced time in the snapshot summary panel.

We have some work coming down the pike to reorganize snapshot summary information in a more logical way within our UI, but for now this surfaces helpful information about the freshness of DUOS Firecloud groups.

Note: we're using the moment library to parse the timestamp value, ex. "an hour ago".  It might be a bit confusing to the user to see a moment on a tab that hasn't been refreshed in awhile, since a snapshot in the UI doesn't get updated periodically in the background (only on refresh or as part of action handling).  But this is consistent with our other moment usages, so I think it's okay.

**Demonstration**

My [developer environment](https://jade-ok.datarepo-dev.broadinstitute.org/snapshots/efb89ba7-8c64-4a7e-ae7a-31faea93d115) is up to date with these changes.

https://user-images.githubusercontent.com/79769153/235708277-de3b59d9-2d77-4d74-b972-1bb47938a892.mov

okotsopoulos committed:
[DR-2969] Display DUOS group's last synced time

If a snapshot is linked to a DUOS dataset, we now display the last synced time of its Firecloud group in the Snapshot Summary panel.
If the Firecloud group has never been synced (shouldn't happen as we sync on group creation), we indicate as much.
Expanded unit tests for these cases.

okotsopoulos committed:
Drive-by: use default null placeholder of (empty) in EditableFieldView

We previously fell back to (Empty) when an EditableFieldValue was missing.
This made for inconsistent behavior across the app.

okotsopoulos commented on src/components/snapshot/overview/SnapshotOverviewPanel.tsx:
```diff
@@ -111,20 +111,30 @@ function SnapshotOverviewPanel(props: SnapshotOverviewPanelProps) {
+                <Typography data-cy="snapshot-duos-last-synced">
+                  {snapshot.duosFirecloudGroup.lastSynced
+                    ? moment(snapshot.duosFirecloudGroup.lastSynced).fromNow()
+                    : 'Never'}
```
We shouldn't find ourselves here: when we create a Firecloud group, we sync its contents.  But figured it was an easy safeguard to put in just in case.

snf2ye commented on src/components/snapshot/overview/SnapshotOverviewPanel.test.tsx:
```diff
@@ -69,26 +107,29 @@ const initialState = {
   },
 };
 
+function mockSnapshotOverviewPanel(snapshot: any): any {
```
Just making sure I understand - you're creating this function instead of using the "BeforeEach" method so that you can pass in a different snapshot for a given test, right?

okotsopoulos replied:
Exactly!  But I am open to a different way to achieve this.

nmalfroy commented on src/components/snapshot/overview/SnapshotOverviewPanel.test.tsx:
```diff
@@ -14,26 +14,64 @@ const duosFirecloudGroup = {
-          { cloudResource: 'firestore', region: 'us-east4' },
-          { cloudResource: 'bucket', region: 'us-east4' },
-        ],
+function createSnapshot(_duosFirecloudGroup?: any): any {
```
I don't know if you want the `_` here right? I think that this indicates that it's an ignored field but you're using it below.  Also, I think there's a `DuosFirecloudGroupModel` type...not sure if it's the same as what's passed in here

okotsopoulos replied:
Oh, good points… I put the underscore because I had a naming conflict with a higher-level `duosFirecloudGroup` object, but I forgot that it is meant to represent an ignored field.  At any rate the conflict probably points to a code smell.

And yes, I think I can use that type here!  I might be able to use more types throughout, now that you mention it.

okotsopoulos replied:
Glad you directed me back here -- I found a near-duplicate block of test set-up code that I could refactor out :)

nmalfroy reviewed (approved):
Looks great!  I have the one question about the test code/typescript stuff

okotsopoulos committed:
Refactor SnapshotOverviewPanel.test

Used generated types where able - DuosFirecloudGroupModel, DatasetModel, SnapshotModel
Previously I prefixed a method input with an underscore to get around a naming conflict.
But this convention indicates an unused input, and the conflict itself pointed to a code smell.
Also cleaned up a near-duplicate block of test set-up code.
